compiler/rustc_lint/src/map_unit_fn.rs

```diff
@@ -0,0 +1,121 @@
+use crate::lints::MappingToUnit;
+use crate::{LateContext, LateLintPass, LintContext};
+
+use rustc_hir::{Closure, ExprKind, QPath, Stmt, StmtKind};
+use rustc_middle::query::Key;
+use rustc_middle::ty;
+
+declare_lint! {
+    /// The `mapping_to_unit` lint checks for `Iterator::map` receive
+    /// a callable that returns `()`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn foo(items: &mut Vec<u8>) {
+    ///     items.sort();
+    /// }
+    ///
+    /// fn main() {
+    ///     let mut x: Vec<Vec<u8>> = vec![
+    ///         vec![0, 2, 1],
+    ///         vec![5, 4, 3],
+    ///     ];
+    ///     x.iter_mut().map(foo);
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Mapping to `()` is almost always a mistake.
+    pub MAP_UNIT_FN,
+    Warn,
+    "`Iterator::map` call that discard the iterator's values"
+}
+
+declare_lint_pass!(MapUnitFn => [MAP_UNIT_FN]);
+
+impl<'tcx> LateLintPass<'tcx> for MapUnitFn {
+    fn check_stmt(&mut self, cx: &LateContext<'tcx>, stmt: &Stmt<'_>) {
+        if stmt.span.from_expansion() {
+            return;
+        }
+
+        if let StmtKind::Semi(expr) = stmt.kind {
+            if let ExprKind::MethodCall(path, receiver, args, span) = expr.kind {
+                if path.ident.name.as_str() == "map" {
+                    if receiver.span.from_expansion()
+                        || args.iter().any(|e| e.span.from_expansion())
+                        || !is_iter(&receiver.kind)
+                    {
+                        return;
+                    }
+                    let ty = cx.typeck_results().expr_ty(&args[0]);
+                    let mut fn_name = "";
+                    if let ExprKind::Path(qpath) = &args[0].kind {
+                        if let QPath::Resolved(_, path) = qpath {
+                            fn_name = path.segments[0].ident.name.as_str();
+                        }
+                    }
+                    if let ty::FnDef(id, _) = ty.kind() {
+                        if let Some(fn_ty) = cx.tcx.fn_sig(id).no_bound_vars() {
```

Comment:
Why `no_bound_vars`? What about fns that take reference arguments with late-bound lifetimes?